torch/_dynamo/output_graph.py

```diff
@@ -2229,6 +2199,296 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
             return arg
         return self.lift_tracked_freevar_to_input(arg)
 
+    # See NOTE: [Auto lift basic free symbols when create_graph_input] for overall design
+    # This function should be used to track the unbacked symbols with its proxy created during
+    # dynamo tracing. It's invoked in wrap_fx_proxy for every proxied call for tensors and symbols.
+    # So that subgraph knows how to bind a symbol input with parent's proxy.
+    def lazy_bind_unbacked_symbols(self, example_value, e_proxy: torch.fx.Proxy):
+        # When binding the symbols in an exmaple_value, we bind the symbols
+        # to the proxy's associatied Tracer instead of current tracer.
+        # This is because:
+        # 1. We maybe calling wrap_tensors during speculate_subgraph because
+        # the variables are lazily realized. The proxy are top-level phs but
+        # current tracer is a subtracer.
+        # 2. For autograd.Function, we trace the backward graph with a new tracer
+        # whose parent is the forward tracer, but we're using all the proxies created
+        # in forward tracer to trace the backward.
+        # For example, forward calls save_for_backward for a input tensor t.
+        # Backward calls t.tolist(). In this case, all the proxies that backward tracer
+        # sees are from parent tracer (i.e. the forward tracer). (e.g. t[0].item())
+        # See test_validate_outputs_unbacked for repro on 2.
+        assert isinstance(e_proxy.tracer, SubgraphTracer)
+        if isinstance(example_value, torch.Tensor):
+            self._lazy_bind_unbacked_symbols_in_tensors(
+                e_proxy.tracer, example_value, e_proxy
+            )
+        elif isinstance(example_value, torch.SymInt):
+            e_proxy.tracer.bound_symbols[example_value.node.expr] = e_proxy
+            set_example_value(e_proxy.node, example_value)
```

Comment:
Why does the "lazy_bind_unbacked_symbols" function have the side effect of setting the example value? I would expect that something else takes care of this.

Reply:
Replied above. The relevant part is: we need to modify bound_symbols in parent (e.g. output of tolist) so that when subgraph looks for proxy of u0 in parent, it can find it in parent's bound_symbols.

Reply:
can we tighten our invariants here? "lazy_bind_unbacked_symbols" should not be tracking backed symints here, so we should check that example_value is unbacked